API for listing index file sizes

camilojd opened the pull request:
I'd like to propose this API to query index file sizes, as part of SegmentsStats.

*Copypasting a comment previously posted on #16131:*

Some comments/questions follow.

I'm consolidating file sizes using the same criteria of `SegmentsStats` to expose disk usage of Lucene index files: `Terms`, `TermVectors`, `StoredFields`, `Norms`, `DocValues` (class `IndexResources` in `SegmentsStats`).

`SegmentsStats` currently exposes memory consumption as bytes of each of the aforementioned so it's possible to add these as part of a class (`IndexResources`) and expose it as two fields of `SegmentsStats`, one for memory resources (existing stats) and other for disk resources. This is what I implemented in my branch, but I'm not sure if it would be better to expose this as another "sister" class of `SegmentsStats`, member of `CommonStats`, and leave `SegmentsStats` as it is. I'm however aware that my `toXContent` serialization could break clients expecting certain fields that now are placed elsewhere.

Some other couple questions I've got:

1. Is it necessary to consolidate sizes by the same criteria of `SegmentsStats` or may be better doing it using the file extensions? This last option has the advantage that would include additional information about postings, i.e., separate size info for positions, payloads, etc.

2. The meat of the matter: Is the way I open the Directory from the `SegmentInfo` and the compound reader a sensible approach? I'm not sure if this could introduce issues with the `Store`.

clintongormley commented:
@mikemccand would you mind having a look at this PR please?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -412,12 +416,69 @@ public final SegmentsStats segmentsStats() {
+                            diskResources.addNormsInBytes(length);
+                        } else if ("tvd".equals(extension) || "tvx".equals(extension) || "tvf".equals(extension)) {
+                            diskResources.addTermVectorsInBytes(length);
+                        }
```
Hmm should we have an "other" category in a final `else` clause here?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:
```diff
@@ -266,17 +319,16 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
     @Override
     public void readFrom(StreamInput in) throws IOException {
         count = in.readVLong();
```
I think we need to check the incoming version and have back-compat here, i.e. if the version is < X (X = first version when we add this feature), we read the old (only memory) stats.

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:
```diff
@@ -287,11 +339,8 @@ public void readFrom(StreamInput in) throws IOException {
     public void writeTo(StreamOutput out) throws IOException {
         out.writeVLong(count);
```
Likewise here.

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -412,12 +416,69 @@ public final SegmentsStats segmentsStats() {
+                }
+                
+                try {
+                    String[] files = directory.listAll();
```
`listAll` and pulling `fileLength` can be costly operations (much more so than pulling memory stats) especially when IO is stressed  ... I wonder if we should make this (pull disk stats in addition to memory stats) an option on the stats request?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -412,12 +416,69 @@ public final SegmentsStats segmentsStats() {
+                        }
+                    }
+                } catch (IOException e) {
+                    // TODO maybe catch specific exceptions for gone files?
```
It's expected that some files will sometimes be gone here, because `IndexWriter` is actively changing the index while this reader has a point-in-time snapshot open.  Maybe we catch `FileNotFoundException` and `NoSuchFileException` and don't log (even trace) those ones?  Any other `IOException` we should log.

mikemccand replied:
Actually, I think IW should not remove files that are being used by NRT readers (we fixed this a while back in Lucene), so I think we should continue to log the exception, but can you increase log level to WARN?

Also,  can you shrink wrap the exception handling?  Put one try/finally around the `listAll`, and another on the `directory.fileLength`?  And include in the exception message which directory (`directory.toString()`) and which file caused an exception.

mikemccand commented:
> Is the way I open the Directory from the SegmentInfo and the compound reader a sensible approach?

Yes, that's correct!

> Is it necessary to consolidate sizes by the same criteria of SegmentsStats or may be better doing it using the file extensions? This last option has the advantage that would include additional information about postings, i.e., separate size info for positions, payloads, etc.

I think I'd rather see by file extension?  E.g. knowing whether your terms dict is huge, or your docs or positions or payloads are huge, is important information.

We could do extensions plus an "interpretation" e.g. ".tim (terms)", ".tip (terms index)", etc.?  Doing it this way also has the advantage that an unknown extension (which can easily happen when lucene changes its file format, e.g. the upcoming dimensional points is a change in 6.0) can just be added into the stats.

camilojd commented:
> We could do extensions plus an "interpretation" e.g. ".tim (terms)", ".tip (terms index)", etc.? Doing it this way also has the advantage that an unknown extension (which can easily happen when lucene changes its file format, e.g. the upcoming dimensional points is a change in 6.0) can just be added into the stats.

I like this compromise. I'll simplify some things and proceed to fix the mentioned issues. :+1:

camilojd commented:
@mikemccand Hi Mike,

I fixed all of the above, and some more. Details:

* Simplified field: just a `ImmutableOpenMap<String, Long>` to store per extension sizes.
* Changed exception handling to limit try/catch blocks to the minimum scope, logging with warn when necessary
* *Transport* serialization doesn't serialize the map field until ES Version 3.0
* XContent serialization takes care of handling known/unknown extension descriptions by querying a map.
* Querying segment disk usage is optional: I added a boolean flag to `CommonStatsFlags`, triggered by setting a `detailed_segments` parameter in the REST endpoint to query segments disk stats. I don't know if there is a nicer way to do this.

So far it works and passes all the tests.

It seems to me that a better place for this would be under `StoreStats`, because `Store` already queries the effective size of all physical files and caches the result... but had no idea/luck on how to grab a `SegmentInfo` for uncommitted segments (only does show some disk usage after forcing a `_flush`). See https://github.com/camilojd/elasticsearch/commit/df0370ff071d94ba90d6b7eae2b7acf004ea921c

mikemccand commented on core/src/main/java/org/elasticsearch/action/admin/indices/stats/CommonStatsFlags.java:
```diff
@@ -192,6 +208,9 @@ public void readFrom(StreamInput in) throws IOException {
         completionDataFields = in.readStringArray();
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
+            detailedSegmentsStats = in.readBoolean();
+        }
```
Maybe add an else clause here to set `detailedSegmentsStats` to `false` so it's explicit?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -418,6 +425,64 @@ public final SegmentsStats segmentsStats() {
+                        directory = engineConfig.getCodec().compoundFormat().getCompoundReader(segmentReader.directory(), segmentCommitInfo.info, IOContext.READ);
+                    } catch (IOException e) {
+                        logger.warn("Error when opening compound reader for Directory [{}] and SegmentCommitInfo [{}]", e,
+                                    segmentReader.directory(), segmentCommitInfo);
```
How about adding `continue` here (we don't process this segment if we couldn't get the `directory`)?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -418,6 +425,64 @@ public final SegmentsStats segmentsStats() {
+                }
+
+                // Couldn't open compound reader, skip
+                if (directory == null) {
```
... and then change this to and `assert directory != null`?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -418,6 +425,64 @@ public final SegmentsStats segmentsStats() {
                 stats.addNormsMemoryInBytes(guardedRamBytesUsed(segmentReader.getNormsReader()));
                 stats.addDocValuesMemoryInBytes(guardedRamBytesUsed(segmentReader.getDocValuesReader()));
+
+                if (!detailed) {
```
Can you factor out all the new code into a new private method `getFileSizeStats` or something, and then invert this if statement to `if (detailed) {...}`?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -418,6 +425,64 @@ public final SegmentsStats segmentsStats() {
+                try {
+                    files = directory.listAll();
+                } catch (IOException e) {
+                    logger.warn("Couldn't list Directory [{}]", e, directory);
```
Maybe just `return` from `getFileSizeStats` here, and then you don't need to set `files = new String[0]` up front?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -418,6 +425,64 @@ public final SegmentsStats segmentsStats() {
+                for (String file : files) {
+                    String extension = IndexFileNames.getExtension(file);
+                    if (extension == null) {
+                        continue;
```
Hmm did you see a `null` extension?  Can you add a comment explaining?  It's spooky if this happens because it becomes a loop-hole in the stats: what if Lucene adds a gigantic per-segment file with no extension :)

camilojd replied:
Oh, I realize this is a leftover from what I prototyped outside Elasticsearch, using only Lucene. It won't be a problem using `SegmentReader`, so I'm taking this out :)

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -418,6 +425,64 @@ public final SegmentsStats segmentsStats() {
+
+                    try {
+                        long length = directory.fileLength(file);
+                        map.put(extension, length);
```
Can you shrink-wrap this try clause?  (Pull the `map.put` out after it.)

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:
```diff
@@ -281,6 +341,17 @@ public void readFrom(StreamInput in) throws IOException {
+                map.put(key, value);
+            }
+            fileExtDiskUsage = map.build();
+        }
```
Can we add `else` clause to set `fileExtDiskUsage` to empty map?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:
```diff
@@ -42,6 +47,27 @@
+            .fPut("tvd", "Term Vector Documents")
+            .fPut("tvf", "Term Vector Fields")
+            .fPut("liv", "Live Documents")
+            .build();
```
Can you add a comment that this is just "best effort", since Lucene changes its extensions over time?  And that it's relatively harmless if the map is missing an extension, since it just means we don't include the description in the stats we return?

mikemccand commented:
Thanks @camilojd, this is looking very nice.  I just left some minor comments.

> Querying segment disk usage is optional: I added a boolean flag to CommonStatsFlags, triggered by setting a detailed_segments parameter in the REST endpoint to query segments disk stats. I don't know if there is a nicer way to do this.

I think `CommonStatsFlags` is fine for now: I don't know of another place to put it (maybe someone else has a suggestion?).

Do we also need to punch the new boolean option through to e.g. the Java client API (and eventually the other language clients)?

> It seems to me that a better place for this would be under StoreStats, because Store already queries the effective size of all physical files and caches the result... but had no idea/luck on how to grab a SegmentInfo for uncommitted segments (only does show some disk usage after forcing a _flush).

I do agree it would be more natural to have the stats there, e.g. we could just fix store stats to break out the file sizes by extension, and not cause any additional IO load to the filesystem.

Maybe it would be an OK limitation that it'd only list flushed segments?  We could explain that in the docs?

camilojd commented:
@mikemccand just pushed changes that addressed all your comments. I really like how the code became more self-contained. :-)

> Do we also need to punch the new boolean option through to e.g. the Java client API (and eventually the other language clients)?

The option is in the Client API, accessible through `IndicesStatsRequest.detailedSegmentsStats(boolean)`, so also through `NodesStatsRequest.indices`. 

Currently it can be requested through HTTP in `RestIndicesStatsAction` and `RestNodesStatsAction`, but it's kinda ugly, just checks if the key `detailed_segments` is in the query string. Maybe leave the option only in the Client API by now?

> Maybe it would be an OK limitation that it'd only list flushed segments? We could explain that in the docs?

It's possible to do that, but it's kind of weird. Currently `StoreStats` already informs the *real* disk usage of the directory, but numbers in my implementation will almost always be very far from the total, because of unflushed segments. `SegmentsStats` is more accurate (not exactly 100% accurate because of cfs, etc).

mikemccand commented:
Thanks @camilojd, I'll review your latest changes.

> The option is in the Client API, accessible through IndicesStatsRequest.detailedSegmentsStats(boolean), so also through NodesStatsRequest.indices.

Oh good, sorry I missed this.

> Currently it can be requested through HTTP in RestIndicesStatsAction and RestNodesStatsAction, but it's kinda ugly, just checks if the key detailed_segments is in the query string. Maybe leave the option only in the Client API by now?

I think it's good to expose the option in the REST API too.

> SegmentsStats is more accurate (not exactly 100% accurate because of cfs, etc).

OK let's leave it where it is now, but maybe add a TODO about whether this could/should move to `StoreStats` instead, explaining the tradeoffs we discussed here?

mikemccand commented on core/src/main/java/org/elasticsearch/action/admin/indices/stats/CommonStatsFlags.java:
```diff
@@ -137,6 +141,15 @@ public CommonStatsFlags completionDataFields(String... completionDataFields) {
         return this.completionDataFields;
     }
 
+    public CommonStatsFlags detailedSegmentsStats(boolean detailedSegmentsStats) {
```
Hmm can we rename `detailedSegmentStats`?  Maybe `includeSegmentFileSizes`?

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -418,12 +425,68 @@ public final SegmentsStats segmentsStats() {
+                logger.warn("Tried to query fileLength but file is gone [{}] [{}]", e, directory, file);
+            } catch (IOException e) {
+                logger.warn("Error when trying to query fileLength [{}] [{}]", e, directory, file);
+            }
```
Instead of inserting 0L into the map, can we just `continue` here?

mikemccand commented:
This change looks great!

I just left a minor (naming, the hardest part!) comment, and let's add a TODO about maybe moving this to `StoreStats` in the future?

Can you add some tests here to confirm the feature is working and catch us if anyone breaks it in the future, and also update the docs explaining this new cool parameter?  Thanks @camilojd!

camilojd commented:
@mikemccand all comments were addressed.

Do you think there is anything left to improve?
Thanks for the amazing help!

mikemccand commented on core/src/main/java/org/elasticsearch/rest/action/admin/indices/stats/RestIndicesStatsAction.java:
```diff
@@ -104,6 +104,10 @@ public void handleRequest(final RestRequest request, final RestChannel channel,
             indicesStatsRequest.fieldDataFields(request.paramAsStringArray("fielddata_fields", request.paramAsStringArray("fields", Strings.EMPTY_ARRAY)));
         }
 
+        if (indicesStatsRequest.segments() && request.hasParam("include_segment_file_sizes")) {
```
Can you update `docs/reference/indices/stats.asciidoc` with this new parameter?

mikemccand commented:
I think we just need to document the new parameter and then we are done!  Thanks @camilojd.

camilojd commented:
@mikemccand now it's done! :+1: Thank you sir!

mikemccand commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -418,12 +425,72 @@ public final SegmentsStats segmentsStats() {
+
+        String[] files;
+        try {
+            files = directory.listAll();
```
Hmm I just realized we are doing too much work here?  We should only be visiting the files belonging to this segment, but we are instead visiting all files in the directory (at least, in the non-compound-file case)?

I think instead of `directory.listAll()`, you could use `segmentReader.getSegmentInfo().files()`?  This should return all file names that this segment uses ...

camilojd replied:
> I think instead of directory.listAll(), you could use segmentReader.getSegmentInfo().files()? This should return all file names that this segment uses ...

Hi @mikemccand,
Makes sense. Perhaps we can do:

            if (useCompoundFile) {
                files = directory.listAll();
            } else {
                files = segmentReader.getSegmentInfo().files().toArray(new String[]{});
            }
Otherwise `getSegmentInfo.files()` returns `.cfs`, `.cfe` files and the Compound Directory can't find them when querying `fileLength`

mikemccand replied:
Ahh yes @camilojd I agree that makes sense!

mikemccand commented:
Thanks @camilojd I left one more comment!

camilojd commented:
@mikemccand I updated the code according to the comments :-) thanks!

mikemccand commented:
Thanks @camilojd, this looks great, I'll push to 5.0 now!

camilojd commented:
Awesome! thank you very much @mikemccand!!
